beam-lib/src/http_util.rs

```diff
@@ -1,21 +1,244 @@
-use http::{request, Request, header};
-use serde::Serialize;
+use std::time::Duration;
 
-use crate::AddressingId;
+use reqwest::{Client, header::{self, HeaderValue}, Url, StatusCode};
+use serde::{de::DeserializeOwned, Serialize};
+use thiserror::Error;
 
+use crate::{AddressingId, TaskRequest, MsgId, TaskResult};
+#[cfg(feature = "sockets")]
+use crate::SocketTask;
 
-pub trait BeamRequestBuilderExt {
-    fn as_app(self, app: &AddressingId, key: &str) -> Self;
-    // We do a generic B here to not require hyper as a dependency
-    fn with_json<B: From<Vec<u8>>, T: Serialize>(self, json: &T) -> Result<Request<B>, http::Error>;
+/// A client used for communicating with the beam network
+#[derive(Debug, Clone)]
+pub struct BeamClient {
+    client: Client,
+    beam_proxy_url: Url,
 }
 
-impl BeamRequestBuilderExt for request::Builder {
-    fn as_app(self, app: &AddressingId, key: &str) -> Self {
-        self.header(header::AUTHORIZATION, format!("ApiKey {app} {key}"))
+#[derive(Error, Debug)]
+pub enum BeamError {
+    #[error("Communication with beam proxy failed: {0}")]
+    ReqwestError(#[from] reqwest::Error),
+    #[error("Unexpected status code {0}")]
+    UnexpectedStatus(StatusCode)
+}
+
+pub type Result<T> = std::result::Result<T, BeamError>;
+
+/// Long polling blocking options
+/// 
+/// The behavior works as follows:
+/// - `wait_count` and `wait_time` are unset => Don't wait
+/// - `wait_count` is set and `wait_time` is unset => Wait for this many items to be ready
+/// - `wait_count` is unset and `wait_time` is set => Wait for the given duration
+/// - `wait_count` is set and `wait_time` is set => Stop waiting after either of the conditions is fulfilled
+#[derive(Debug)]
+pub struct BlockingOptions {
+    /// How long to poll for
+    pub wait_time: Option<Duration>,
+    /// The number of elements to wait for
+    pub wait_count: Option<u16>,
```

Comment:
Is there a specific reason for choosing `u16` instead of `usize`?

Reply:
It is u16 in beam internally as well